### DISCUSSION TITLE
minted结构体在正文中如何引用？

### DISCUSSION THREAD

**Discussion by Auyuir** [tuna/thuthesis#877] [Q&A] [answered]
注意到cls模板中添加了对minted的支持，可以把listing转译为“代码”
请问如何在正文中引用minted结构体，下图的写法中使用\ref引用minted结构体所在的listing结构体时，生成的pdf无法正确识别。
![image](https://github.com/tuna/thuthesis/assets/31274112/1350df97-6680-48d4-97d3-c2f8ed266bd1)
生成的pdf：
![image](https://github.com/tuna/thuthesis/assets/31274112/2b099944-65c2-4fc2-b9d7-860412396c2d)

**Comment by atxy-blip** [comment, marked as answer]
没能用下面的例子复现问题，建议重新编译一遍看看。

```tex
\documentclass{thuthesis}
\usepackage{minted}

\begin{document}

如代码~\ref{C++top}~所示

\begin{listing}[H]
\mint{C++}/int i = 1;/ 
\caption{Example of a listing.}
\label{C++top}
\end{listing}

\end{document}
```

**Reply by Auyuir** [reply]
谢谢，清理辅助文件+重新编译后问题已解决。

额外请教一下“代码”程序块的caption序号否可以携带章节号？
比如原提问中生成的序号是1而不是带有章节号的5.1

**Reply by note286** [reply]
可以，重定义一下即可，不过可能有的导师会要求将代码视作图，即，放在`figure`浮动体中，不过你确认导师允许，使用代码环境也可以。

可以参考 [formatting - Renewing the \thelstlisting command when hyperref is on - TeX - LaTeX Stack Exchange](https://tex.stackexchange.com/questions/261437/renewing-the-thelstlisting-command-when-hyperref-is-on) 。

**Reply by note286** [reply]
类似这种:
```latex
\renewcommand{\thelisting}{\arabic{chapter}.\arabic{listing}}
```

**Reply by atxy-blip** [reply]
更简洁的方法是在导言区加 `\counterwithin{listing}{chapter}`。

**Reply by note286** [reply]
我只是担心没准他又要`5-1`这种的了，毕竟`\counterwithin`里的`.`是写死的。

**Reply by atxy-blip** [reply]
`\counterwithin` 其实还会在每章开始时重置 `listing` 计数器。要改连接符还是拿 `etoolbox` 做 patch 吧（

**Reply by Auyuir** [reply]
请问使用
```
\renewcommand{\thelisting}{\arabic{\thechapter}.\arabic{listing}}
```
和cls模板中的
```
\renewcommand\thefigure{%
  \ifnum\c@chapter>\z@
    \thechapter
    \thu@figure@number@separator
  \fi
  \@arabic\c@figure
}
```
有哪些区别？我用前者编译时会发生编译错误，显示
```
Missing number, treated as zero. \caption{。。。}
```
生成的半成品PDF中，正文引用和代码题注都显示0.1

**Reply by note286** [reply]
不好意思，我疏忽了。

```latex
\makeatletter
\@addtoreset{listing}{chapter}
\makeatother
\renewcommand{\thelisting}{\arabic{chapter}.\arabic{listing}}
```

**Reply by note286** [reply]
或者
```latex
\counterwithin*{listing}{chapter}
\renewcommand{\thelisting}{\arabic{chapter}.\arabic{listing}}
```
也是可以的，`\counterwithin*`与`\@addtoreset`是等价的。

如果只需要分隔符为`.`，那么`\counterwithin`就足够了，`\counterwithin`里的`.`是写死的。

**Reply by note286** [reply]
> 请问使用
> 
> ```
> \renewcommand{\thelisting}{\arabic{\thechapter}.\arabic{listing}}
> ```
> 
> 和cls模板中的
> 
> ```
> \renewcommand\thefigure{%
>   \ifnum\c@chapter>\z@
>     \thechapter
>     \thu@figure@number@separator
>   \fi
>   \@arabic\c@figure
> }
> ```
> 
> 有哪些区别？我用前者编译时会发生编译错误，显示


这里这么定义，是先判断`chapter`计数器是否大于0，如果不大于0，就仅输出figure，这样可以避免有人不写`chapter`，出现`0.1`这种情况，此时就只会显示`1`。当然，学位论文不可能不写章，所以实际使用的时候，不会发生这种情况。

**Reply by Auyuir** [reply]
非常感谢！

**Comment by Auyuir** [comment]
```\counterwithin{listing}{chapter}```
能够生效，谢谢！
